### PR TITLE
DPKG algorithm improvement

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
|#22442|

## Objective
This PR aims to bolster the DPKG matcher's robustness and reliability by incorporating validation checks for version and revision inputs, mirroring the logic found in the original `dpkg` library. The enhancement is intended to identify and warn about invalid inputs before processing, ensuring that the matcher does not return or rely on invalid version information.

## Description
- In our ongoing efforts to align closely with the best practices and standards established by the Debian package management system, we identified a gap in our DPKG matcher's handling of version and revision inputs. Specifically, the matcher previously lacked the capability to rigorously validate these inputs against the criteria used by the `dpkg` library, potentially leading to the acceptance of malformed versions and subsequent inaccuracies in package matching.
- To address this, we have updated the DPKG matcher to avoid returning false positives when a version or release part has invalid characters.

### Static Analysis
- Comprehensive static analysis was performed on the updated code to ensure that it meets our high standards for quality, efficiency, and security. This step also helped in identifying areas where the implementation could be further optimized.

#### Unit Testing
- A series of unit tests were refactor to rigorously test the new validation logic within the DPKG matcher. These tests cover a wide range of input scenarios, including both valid and invalid versions and revision strings, to ensure that the matcher accurately identifies and warns about problematic inputs.

#### Integration Testing
- Integration testing was conducted to assess the impact of the new validation checks on the overall package management workflow. This included testing the matcher's interaction with other system components and verifying that the addition of input validation does not adversely affect system performance or usability.
- **Testing Steps**:
  1. Install manager
  2. Connect agents with `deb` format.

<details><summary>Report</summary>

image
Summary dashboard vulnerabilities

image
Details of vulnerabilities

Some vulnerabilities are now found, for example:

- CVE-2024-25062
- CVE-2023-27043

The warning it doesn't mean that the comparative can't be released:

```bash
root@agent_debian_12:/# dpkg --compare-versions 3.5_alpha lt 3.7_alpha && echo 1
dpkg: warning: version '3.5_alpha' has bad syntax: invalid character in version number
dpkg: warning: version '3.7_alpha' has bad syntax: invalid character in version number
1
```

</details>

## Aditional notes

The vulnerability  CVE-2002-2436, and others related with Firefox, has invalid elements in content

```text
DEBUG: Scanning package - 'firefox' (Installed Version: 116.0.2-1, Security Vulnerability: CVE-2002-2436). Identified vulnerability: Version: 3.0.4. Required Version Threshold: . Required Version Threshold (or Equal): .
DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 116.0.2-1
DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 3.0.4

DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (116.0.2-1 vs 52.0.).
```

### CODE REVIEW

**Comment by GabrielEValenzuela** [comment]
We found that `snap` packages follow a versioning similar to `dpkg` so, we define the DPKG strategy for these packages.

Here is an example of installed packages in `AlmaLinux 9.3`. 

image

This change is validated in the new steps in `TC-007`

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 1]
```diff
@@ -217,25 +232,41 @@ class VersionObjectDpkg final : public IVersionObject
...
+        }
         for (; *ptr; ptr++)
         {
             if (!std::isdigit(*ptr) && !std::isalpha(*ptr) && strchr(".-+~:", *ptr) == NULL)
-                return false;
+            {
+                logWarn(WM_VULNSCAN_LOGTAG,
+                        "Invalid character in revision: %s in %s.",
+                        output.version.c_str(),
+                        version.c_str());
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:258]: Lets move this to debug1

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 2]
```diff
@@ -217,25 +232,41 @@ class VersionObjectDpkg final : public IVersionObject
...
         }
         for (ptr = output.revision.c_str(); *ptr; ptr++)
         {
             if (!std::isdigit(*ptr) && !std::isalpha(*ptr) && strchr(".+~", *ptr) == NULL)
-                return false;
+            {
+                logWarn(WM_VULNSCAN_LOGTAG,
+                        "Invalid character in revision: %s in %s.",
+                        output.revision.c_str(),
+                        version.c_str());
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:268]: Lets move this to debug1

File: src/wazuh_modules/vulnerability_scanner/testtool/versionMatcher/version_verifier.sh
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/testtool/versionMatcher/version_verifier.sh:1]: Thanks for updating the testtool!

**Review by sebasfalcone — COMMENTED** [review]
- Code 🟢  
  - Looks clean and fixes coding style misalignments 
  - It adds more cases to the testtool! 🥳

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Done! :)

**Comment by Dwordcito** [comment]
Reason to delete tests 007 ?

**Comment by sebasfalcone** [comment]
> Reason to delete tests 007?
Same question

**Review by sebasfalcone — COMMENTED** [review]
LGTM! We just need to address the removal of the 007 test

- Code 🟢 
- Test 🟢 

Installed firefox via snap
image
image

Only errors associated:
```
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs preview_release).'
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 
'Unable to compare versions (123.0.1-1 vs preview_release).'
```
- Addressed here
  - https://github.com/wazuh/intelligence-data/issues/70

```
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:146 at operator()(): DEBUG: Scanning package - 'firefox' (Installed Version: 123.0.1-1, Security Vulnerability: CVE-2012-0463). Identified vulnerability: Version: (4.0). Required Version Threshold: . Required Version Threshold (or Equal): 10.0.
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] versionMatcher.hpp:195 at createVersionObject(): DEBUG: Error creating VersionObject (DPKG). Version string doesn't match the specified type. Version string: (4.0)
2024/03/15 10:38:59 wazuh-modulesd:vulnerability-scanner[40224] packageScanner.hpp:310 at operator()(): DEBUG: Failed to scan package: 'firefox', CVE Numbering Authorities (CNA): 'nvd', Error: 'Unable to compare versions (123.0.1-1 vs (4.0)).'
```
- Firefox has a [weird versioning convention](https://es.wikipedia.org/wiki/Anexo:Versiones_de_Firefox), every mayor release they remove the minor number. This generates some problems on the validation (@Damian-Mangold I want to synchronize this)

**Comment by GabrielEValenzuela** [comment]
> > Reason to delete tests 007?
> 
> Same question

Restored! Sorry!

**Review by sebasfalcone — APPROVED** [review]

**Pull request merged**